### DISCUSSION TITLE
IGA Negative Jacobian

### DISCUSSION THREAD

**Discussion by makeclean** [idaholab/moose#22248] [Q&A General] [answered]
Hi All

Quick one, following successful recompilation of MOOSE with an appropriate derivative size (:facepalm:) my original IGA problem ran, I then remeshing to remove some aberrations from the geometry, the hex mesh underlying the geometry is good, and produces an IGA mesh without issue, but when MOOSE runs this new geometry I get a -ve jacobian warning, I've traced the IGA element to a particular hex element but it doesn't have a -ve jacobian, do you have any advice on how to trace down the issue, and fix the upstream mesh?

Thanks

Andy

**Comment by GiudGiud** [comment, marked as answer]
Hello

There is an offset of 1 between the ids, can you double check you traced it back to the right element?
Another technique would be to use a VolumeAux to output the element volumes, then use a filter on that number in paraview to find the bad one

Guillaume

**Reply by makeclean** [reply]
So, the element suggested by the bounding box of the hex is shown below, I dont see how this could have a -ve Jacobian. The only issue might be that the error regarding the -ve jacobian happens during loading the mesh, not during running. 

![image](https://user-images.githubusercontent.com/2439754/193268171-0c66d3a2-d407-4cef-bc0f-0b6bd528ea9d.png)

**Reply by makeclean** [reply]
Here is the text
```
  Finished Initializing                                                                  [
  3.25 s] [  716 MB]
Finished Setting Up                                                                      [
  5.03 s] [  885 MB]
Framework Information:
MOOSE Version:           git commit 82f5b1d92b on 2022-09-26
LibMesh Version:         1eef758792bff86e26258972ab9dc60533161964
PETSc Version:           3.16.6
SLEPc Version:           3.16.2
Current Time:            Fri Sep 30 13:43:48 2022
Executable Timestamp:    Tue Sep 27 16:01:28 2022

Parallelism:
  Num Processors:          4
  Num Threads:             1

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   
    Total:                 149904
    Local:                 37495
    Min/Max/Avg:           36828/38004/37476
  Elems:                   
    Total:                 16896
    Local:                 4224
    Min/Max/Avg:           4224/4224/4224
  Num Subdomains:          3
  Num Partitions:          4
  Partitioner:             metis

Nonlinear System:
  Num DOFs:                599616
  Num Local DOFs:          149980
  Variables:               { "temperature" "disp_x" "disp_y" "disp_z" } 
  Finite Element Types:    "RATIONAL_BERNSTEIN" 
  Approximation Orders:    "SECOND" 

Auxiliary System:
  Num DOFs:                16896
  Num Local DOFs:          4224
  Variables:               "volume" 
  Finite Element Types:    "MONOMIAL" 
  Approximation Orders:    "CONSTANT" 

Execution Information:
  Executioner:             Steady
  Solver Mode:             NEWTON
  MOOSE Preconditioner:    SMP (auto)

  Elem Information
   id()=52, unique_id()=78286, processor_id()=3
   type()=HEX27
   dim()=3
   n_nodes()=27
   mapping=RATIONAL_BERNSTEIN_MAP
    0  Node id()=954, processor_id()=3, Point=(x,y,z)=(   0.001, -0.000293901, 0.00940452)
    DoFs=(0/0/449088) (0/1/449089) (0/2/449090) (0/3/449091) 
    weight=1
    1  Node id()=833, processor_id()=3, Point=(x,y,z)=(0.00101849, -0.000546374, 0.0094420
3)
    DoFs=(0/0/448736) (0/1/448737) (0/2/448738) (0/3/448739) 
    weight=1
    2  Node id()=834, processor_id()=3, Point=(x,y,z)=( 0.00025, -0.000558127, 0.00944398)
    DoFs=(0/0/448740) (0/1/448741) (0/2/448742) (0/3/448743) 
    weight=1
    3  Node id()=955, processor_id()=3, Point=(x,y,z)=( 0.00025, -0.000293901, 0.00940452)
    DoFs=(0/0/449092) (0/1/449093) (0/2/449094) (0/3/449095) 
    weight=1
    4  Node id()=930, processor_id()=3, Point=(x,y,z)=(   0.001, -0.00040452, 0.0092939)
    DoFs=(0/0/449028) (0/1/449029) (0/2/449030) (0/3/449031) 
    weight=1
    5  Node id()=855, processor_id()=3, Point=(x,y,z)=(0.000984093, -0.000715071, 0.009343
99)
    DoFs=(0/0/448824) (0/1/448825) (0/2/448826) (0/3/448827) 
    weight=1
    6  Node id()=856, processor_id()=3, Point=(x,y,z)=( 0.00025, -0.000717885, 0.00934319)
    DoFs=(0/0/448828) (0/1/448829) (0/2/448830) (0/3/448831) 
    weight=1
    7  Node id()=931, processor_id()=3, Point=(x,y,z)=( 0.00025, -0.00040452, 0.0092939)
    DoFs=(0/0/449032) (0/1/449033) (0/2/449034) (0/3/449035) 
    weight=1
    8  Node id()=956, processor_id()=3, Point=(x,y,z)=(0.00102408, -0.000302228, 0.0093982
5)
    DoFs=(0/0/449096) (0/1/449097) (0/2/449098) (0/3/449099) 
    weight=1
    9  Node id()=845, processor_id()=3, Point=(x,y,z)=(0.000649218, -0.000528803, 0.009442
47)
    DoFs=(0/0/448784) (0/1/448785) (0/2/448786) (0/3/448787) 
    weight=1
    10  Node id()=957, processor_id()=3, Point=(x,y,z)=( 0.00025, -0.00031551, 0.00939682)
    DoFs=(0/0/449100) (0/1/449101) (0/2/449102) (0/3/449103) 
    weight=1
    11  Node id()=958, processor_id()=3, Point=(x,y,z)=(0.000625, -0.000293901, 0.00940452
)
    DoFs=(0/0/449104) (0/1/449105) (0/2/449106) (0/3/449107) 
    weight=1
    12  Node id()=959, processor_id()=3, Point=(x,y,z)=(   0.001, -0.000357971, 0.00935797
)
    DoFs=(0/0/449108) (0/1/449109) (0/2/449110) (0/3/449111) 
    weight=1
    13  Node id()=859, processor_id()=3, Point=(x,y,z)=(0.000993625, -0.000670314, 0.00939
969)
    DoFs=(0/0/448840) (0/1/448841) (0/2/448842) (0/3/448843) 
    weight=1
    14  Node id()=860, processor_id()=3, Point=(x,y,z)=( 0.00025, -0.000673075, 0.00940009
)
    DoFs=(0/0/448844) (0/1/448845) (0/2/448846) (0/3/448847) 
    weight=1
    15  Node id()=960, processor_id()=3, Point=(x,y,z)=( 0.00025, -0.000357971, 0.00935797
)
    DoFs=(0/0/449112) (0/1/449113) (0/2/449114) (0/3/449115) 
    weight=1
    16  Node id()=936, processor_id()=3, Point=(x,y,z)=(0.000989257, -0.000485761, 0.00929
217)
    DoFs=(0/0/449036) (0/1/449037) (0/2/449038) (0/3/449039) 
    weight=1
    17  Node id()=861, processor_id()=3, Point=(x,y,z)=(0.000608948, -0.000710116, 0.00934
637)
    DoFs=(0/0/448848) (0/1/448849) (0/2/448850) (0/3/448851) 
    weight=1
    18  Node id()=937, processor_id()=3, Point=(x,y,z)=( 0.00025, -0.000496346, 0.00929221
)
    DoFs=(0/0/449040) (0/1/449041) (0/2/449042) (0/3/449043) 
    weight=1
    19  Node id()=938, processor_id()=3, Point=(x,y,z)=(0.000625, -0.00040452, 0.0092939)
    DoFs=(0/0/449044) (0/1/449045) (0/2/449046) (0/3/449047) 
    weight=1
    20  Node id()=961, processor_id()=3, Point=(x,y,z)=(0.000654528, -0.000277344, 0.00940
385)
    DoFs=(0/0/449116) (0/1/449117) (0/2/449118) (0/3/449119) 
    weight=1
    21  Node id()=962, processor_id()=3, Point=(x,y,z)=(0.000999709, -0.000404649, 0.00935
081)
    DoFs=(0/0/449120) (0/1/449121) (0/2/449122) (0/3/449123) 
    weight=1
    22  Node id()=867, processor_id()=3, Point=(x,y,z)=(0.000619607, -0.000668235, 0.00940
113)
    DoFs=(0/0/448872) (0/1/448873) (0/2/448874) (0/3/448875) 
    weight=1
    23  Node id()=963, processor_id()=3, Point=(x,y,z)=( 0.00025, -0.000418708, 0.00935042
)
    DoFs=(0/0/449124) (0/1/449125) (0/2/449126) (0/3/449127) 
    weight=1
    24  Node id()=964, processor_id()=3, Point=(x,y,z)=(0.000625, -0.000357971, 0.00935797
)
    DoFs=(0/0/449128) (0/1/449129) (0/2/449130) (0/3/449131) 
    weight=1
    25  Node id()=947, processor_id()=3, Point=(x,y,z)=(0.000615466, -0.000472034, 0.00929
731)
    DoFs=(0/0/449064) (0/1/449065) (0/2/449066) (0/3/449067) 
    weight=1
    26  Node id()=965, processor_id()=3, Point=(x,y,z)=(0.000628136, -0.000385747, 0.00935
694)
    DoFs=(0/0/449132) (0/1/449133) (0/2/449134) (0/3/449135) 
    weight=1
   n_sides()=6
    neighbor(0)=56
    neighbor(1)=468
    neighbor(2)=46
    neighbor(3)=nullptr
    neighbor(4)=nullptr
    neighbor(5)=51
   hmin()=0.000156439, hmax()=0.000863727
   volume()=2.81264e-11
   active()=1, ancestor()=0, subactive()=0, has_children()=0
   parent()=nullptr
   level()=0, p_level()=0
   refinement_flag()=DO_NOTHING
   p_refinement_flag()=DO_NOTHING
   DoFs=(1/0/12694) 
ERROR: negative Jacobian -6.27634e-14 at point (x,y,z)=(0.000915474, -0.000293901, 0.00940
452) in element 52
[3] ../src/fe/fe_map.C, line 1173, compiled Sep 27 2022 at 15:52:56
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 3 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
```

**Reply by GiudGiud** [reply]
Cubit should support Jacobian calculations for this element according to this
https://coreform.com/cubit_help/mesh_generation/mesh_quality_assessment/hexahedral_metrics.htm
let s see if it agrees

I m going to plot this element and see what I get for these metrics. A volume of 2e-11 is pretty small already but it's positive. 
Negative Jacobian, not sure why that is

let s tag @roystgnr since this is for IGA

**Reply by makeclean** [reply]
nb: i applied a scaling factor of 1e-3 to the mesh wrt to the raw cubit geometry

**Reply by GiudGiud** [reply]
This is what I get so far for the nodes
![nodes](https://user-images.githubusercontent.com/25037151/193311667-3e3ef245-4e8b-4195-93d2-3f3fb45a705f.png)

interesting that some face centers on 4 faces are offset, and so is the center

**Reply by GiudGiud** [reply]
the top right points on the figure could be inverted?
I ll see what I can get for Jacobian

**Reply by GiudGiud** [reply]
projecting the faces of the element onto an XY plane I get the image below
the center node is outside of the face (side element). 

There's another face that is similarly deformed. I imagine two issues can make for a positive volume and a negative Jacobian
The other faces look more like expected

So this mesh is straight from the mesher? no deforming physics solved on it yet?
![face_1](https://user-images.githubusercontent.com/25037151/193359017-04bed87e-9626-4c77-9e74-18eb3b97d038.png)

**Reply by GiudGiud** [reply]
The desired map, modulo a rotation
```
# Map of side element nodes
# *          3     6     2
# *   QUAD9: o-----o-----o
# *          |           |           eta
# *          |     8     |            ^
# *        7 o     o     o 5          |
# *          |           |            |
# *          |           |            o---> xi
# *          o-----o-----o
# *          0     4     1
```

**Reply by makeclean** [reply]
No physics, straight from the cubit uspline process barfs on startup

**Reply by GiudGiud** [reply]
@GregVernon @roystgnr thoughts?

**Reply by roystgnr** [reply]
IIRC Cubit can still require some work to give you invertible mappings if your geometry is tricky enough (very fine fillets were what were killing us on the pressure vessel mesh we were using for initial large-scale testing), but Cubit ought to be able to tell you if it sees a problem in that respect ... maybe?  I'm not sure if there's a theorem for IGA that proves anything like "if the jacobian is invertible at these N points then it's invertible everywhere".

These are all weight=1 nodes, shouldn't be giving any trickiness there.

That projected face looks horrible, but it's important to remember that the non-vertex nodes here are control points, not mid-edge or mid-face nodes, so they "pull" from farther away than the element itself stretches; that could be an entirely reasonable coordinate set.

Wish I had the xi/eta/zeta for that failing coordinate, not just x/y/z.  Could we get that out of a debugger (moving up the stack to FE<Dim,T>::reinit, then printing either `pts[p]` or `qrule->get_points()[p]` depending on which branch you're in)?  That line number is coming from `compute_single_point_map`, so it should be a forward calculation of a quadrature point failing (rather than some physics assuming lagrangian nodal positions and being unpleasantly surprised in an inverse map), but it'd be good to be sure.

I'm trying to think of what else we might do to make this easier to figure out.  Generating edge and face elements to see if any of them come up non-invertible anywhere?  I'm not sure if we have any code to do that which preserves the Bezier-Bernstein mapping type on the new elements.  Evaluating Jacobians at each vertex, where they'd be easier to double-check by hand than at an arbitrary interior quadrature point?

**Reply by GregVernon** [reply]
> That projected face looks horrible, but it's important to remember that the non-vertex nodes here are control points, not mid-edge or mid-face nodes, so they "pull" from farther away than the element itself stretches; that could be an entirely reasonable coordinate set.

Just to add to this, here's a video I made a while ago showing the difference between an interpolatory basis (Lagrange) and a non-interpolatory basis (Bernstein) that demonstrates what @roystgnr is talking about:  https://www.youtube.com/watch?v=Czr4BtTbDb4

**Reply by makeclean** [reply]
With @GregVernon's help from the Cubit side, I can see Cubit doesn't mention anything regarding a negative Jacobian on their side, however the resultant element is quite 'tangled' or 'self-intersecting' could that be what this would be referring to?

![image](https://user-images.githubusercontent.com/2439754/194018588-82c214f0-6b96-4774-8e3b-80ad99fa5ce1.png)

Note, I'm using the `quality uspline 1 scaled jacobian threshold <x>` command to highlight them, all of the ones with a particularly low Jacobian all have this self-intersection.

**Reply by roystgnr** [reply]
If by self-intersecting you mean there are two points on the master element that coincide at the same point in physical space, that definitely implies a zero Jacobian (and if there's a neighborhood around the points, a negative Jacobian) in the mapping.

You can sometimes solve on an element whose mapping isn't bijectve anyway, by limiting where the self-intersection occurs (quads that get turned into triangles by merging vertices are common in some industry codes, and I know there are some libMesh codes doing the same for code-to-code comparison tests) and using a quadrature rule + weak formulation combo designed to keep it stable, but if you're getting self-intersection by accident it's almost guaranteed to be a disaster.

**Comment by makeclean** [comment]
Well I'm meaning (correctly or otherwise) that the IGA element on the right hand figure you can see two curves of the element cross over one another
